### PR TITLE
TIP-159: Remove getCompleteness from models

### PR DESCRIPTION
**Pull request by pierallard**

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/CompleteFilter.php [hunk 1]
```diff
@@ -18,55 +19,52 @@
...
-            ->innerJoin('variant_product.completenesses', 'completeness')
-            ->innerJoin('completeness.locale', 'locale')
-            ->innerJoin('completeness.channel', 'channel')
-            ->setParameter(':product_model', $productModel->getId());
-
-        $result = $queryBuilder->getQuery()->getArrayResult();
-
-        return new CompleteFilterData($result);
+        return new CompleteFilterData(
+            $this->connection->executeQuery($sql, ['product_model_id' => $productModel->getId()])->fetchAll()
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/CompleteFilter.php]: personally, I think inlining is pretty hard to debug

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            PRIMARY KEY (`id`),
+            UNIQUE KEY `searchunique_idx` (`channel_id`,`locale_id`,`product_id`),
+            KEY `IDX_113BA854E559DFD1` (`locale_id`),
+            KEY `IDX_113BA85472F5A1AA` (`channel_id`),
+            KEY `IDX_113BA8544584665A` (`product_id`),
+            KEY `ratio_idx` (`ratio`),
+            CONSTRAINT `FK_113BA8544584665A` FOREIGN KEY (`product_id`) REFERENCES `pim_catalog_product` (`id`) ON DELETE CASCADE,
+            CONSTRAINT `FK_113BA85472F5A1AA` FOREIGN KEY (`channel_id`) REFERENCES `pim_catalog_channel` (`id`) ON DELETE CASCADE,
+            CONSTRAINT `FK_113BA854E559DFD1` FOREIGN KEY (`locale_id`) REFERENCES `pim_catalog_locale` (`id`) ON DELETE CASCADE
+        )";
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php]: You should specify charset, engine, and drop table if exist I think.

```
ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```
Such as:
```
DROP TABLE IF EXISTS akeneo_structure_version_last_update;
CREATE TABLE akeneo_structure_version_last_update (
    resource_name varchar(255) NOT NULL,
    last_update datetime NOT NULL COMMENT '(DC2Type:datetime)',
    PRIMARY KEY(resource_name)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
```

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        $this->connection->exec($completenessTableSql);
+
+        $missingTableSql = "CREATE TABLE `pim_catalog_completeness_missing_attribute` (
+            `completeness_id` int(11) NOT NULL,
+            `missing_attribute_id` int(11) NOT NULL,
+            PRIMARY KEY (`completeness_id`,`missing_attribute_id`),
+            KEY `IDX_DF61BBD3B640FBA5` (`completeness_id`),
+            KEY `IDX_DF61BBD3762147F6` (`missing_attribute_id`),
+            CONSTRAINT `FK_DF61BBD3762147F6` FOREIGN KEY (`missing_attribute_id`) REFERENCES `pim_catalog_attribute` (`id`) ON DELETE CASCADE,
+            CONSTRAINT `FK_DF61BBD3B640FBA5` FOREIGN KEY (`completeness_id`) REFERENCES `pim_catalog_completeness` (`id`) ON DELETE CASCADE
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php]: same

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    public function initDbSchema(InstallerEvent $event): void
+    {
+        $completenessTableSql = "CREATE TABLE `pim_catalog_completeness` (
+            `id` int(11) NOT NULL AUTO_INCREMENT,
+            `locale_id` int(11) NOT NULL,
+            `channel_id` int(11) NOT NULL,
+            `product_id` int(11) NOT NULL,
+            `ratio` int(11) NOT NULL,
```
Comment (anaelChardan) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php]: you can remove this column

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php (hunk 2 above)
Comment (anaelChardan) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/InitCompletenessDbSchemaSubscriber.php]: I do not understand why would you remove a table at install while the table does not exist yet

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductInterface.php [hunk 5]
```diff
@@ -113,22 +113,6 @@ public function setEnabled($enabled);
      */
     public function isEnabled();
 
-    /**
-     * Get product completenesses
-     *
-     * @return Collection of CompletenessInterface
-     */
-    public function getCompletenesses();
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductInterface.php]: :tada:

File: src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/CompleteFilter.php [hunk 6]
```diff
@@ -6,6 +6,7 @@
 use Akeneo\Pim\Enrichment\Component\Product\Model\ProductModelInterface;
 use Akeneo\Pim\Enrichment\Component\Product\ProductAndProductModel\Query\CompleteFilterData;
 use Akeneo\Pim\Enrichment\Component\Product\ProductAndProductModel\Query\CompleteFilterInterface;
+use Doctrine\DBAL\Connection;
 use Doctrine\ORM\EntityManagerInterface;
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Doctrine/ORM/Query/CompleteFilter.php]:
```suggestion
```

**Pull request merged**